### ISSUE TITLE
How can I apply sigmoid activation function in output layer? 

### ISSUE THREAD

**Issue by chhyun** [Microsoft/CNTK#1919]
I'm trying to apply sigmoid function in output layer with using cross-entropy loss function

please take a loot at my code. I'm using ndl file not brainscript.

this is my network below
      ...
     # network composition
      h1 = DNNImageSigmoidLayer(8, 8, cMap2, hiddenDim, pool2, 1)
      d1 = Dropout(h1)
      h2 = DNNSigmoidLayer(hiddenDim, hiddenDim, d1, 1)
      d2 = Dropout(h2)
      output = DNNSigmoidLayer(hiddenDim, labelDim, d2, 1)  
      ce = CrossEntropy(label, out)
      ...

is this right code to apply sigmoid function to output layer?
cause the training was not successful. And when I used Deeplearning Toolbox with matlab,
the training with sigmoid(in output layer) and cross-entropy was successful.    

 one more question : In cntk, does CrossEntropyWithSoftmax apply softmax function to network output value? And is the network output from DNNLayer such as

         output = DNNLayer(hiddenDim, labelDim, d2, 1)  

just a network output value with not applying activation function?

**Comment by cha-zhang** [comment]
Yes, CrossEntropyWithSoftmax will apply both softmax and cross entropy. These two are combined to run faster. Your output will not have softmax and you could do that manually. 

The reason is if you only care about max, you don't have to run softmax during inference.

**Comment by chhyun** [comment]
Thank you for your answer. 

And i gotta ask you few more, cuz it's quite confusing...

Q1. So you mean, DNNSigmoidLayer in the output layer doesn't work, and I should manually 
apply sigmoid function with extracted network output value?

Q2. Why is there no way to automatically apply sigmoid in the output layer? Is cntk a framework developed solely for the use of softmax and cross entropy? All I want to do is just applying sigmoid in the output layer and use cross-entropy loss function for training, and I don't get why I can't do this simple thing.

**Comment by cha-zhang** [comment]
Sorry for the confusion. Could you elaborate what do mean "training was not successful"? 

I see: 
output = DNNSigmoidLayer(hiddenDim, labelDim, d2, 1)
ce = CrossEntropy(label, out)

shouldn't it be:
ce = CrossEntropy(label, output)?

**Comment by chhyun** [comment]
yes 
ce = CrossEntropy(label, output) is right 

"training was not successful" means the loss didn't go down while training.
loss stayed at the same value and didn't changed.
so learning was not done at all.

**Comment by cha-zhang** [comment]
If your network is simple, would you mind giving us a repro? 

There are many reasons that a network doesn't converge. It could be wrong settings of learning rate, or initialization. However, if this is an example in CNTK and simply switching from CrossEntropyWithSoftMax to SoftMax and CrossEntry will cause issue, definitely let us know and we will figure out why.

**Comment by chhyun** [comment]
Okay. 

Let me know your email and what you need. 

This is my email : chhyun@knu.ac.kr

**Comment by spandantiwari** [comment]
I looked at the repro you sent to cha-zhang. The issue may be arising due to confusion between sigmoid and softmax. Roughly speaking, sigmoid can be treated as a special case of softmax for the binary case, and softmax is the generalization to multiple classes. I understand that you have a multi-class (`labelDim` > 2) output.  In the code you shared, you are passing the output of `DNNSigmoidLayer` to `CrossEntropy`. `CrossEntropy ` works with (posterior) probability distributions and the input must sum up to 1. The output of `DNNSigmoidLayer `(it applies sigmoid, not softmax) will not be normalized to sum up to 1.

So, instead of `DNNSigmoidLayer`, maybe you can try just using a linear layer, say `DNNLayer`. If you really need to apply sigmoid you can pass the output to `sigmoid()` separately, and then wire the output of the linear layer `CrossEntropyWithSoftmax `as the criterion.
```

      h3 = DNNLayer(hiddenDim, ageDim, d2, 1)
      ageol = sigmoid(h3); # Only if needed
      ce = CrossEntropyWithSoftmax (age, h3)  

```

**Comment by spandantiwari** [comment]
Chhyun - did this solve the issue?

**Comment by chhyun** [comment]
Thanks for your answer. But why do I have to use CrossEntropyWithSoftmax ? .

Cause I intend to apply sigmoid to output layer, not softmax.

With the code you've sent to me, the output value would be applied both
sigmoid and softmax.

So I tried

      h3 = DNNLayer(hiddenDim, ageDim, d2, 1)
      ageol = sigmoid(h3); # Only if needed
      ce = CrossEntropy (ageol, h3)


But still have the same problem as before.



2017-06-07 5:29 GMT+09:00 Spandan Tiwari <notifications@github.com>:

**Comment by spandantiwari** [comment]
Since you intend to obtain output from sigmoid, `ageol` is just that. To your question about why `softmax `is needed, see my explanation in my answer above:

> I understand that you have a multi-class (labelDim > 2) output. In the code you shared, you are passing the output of DNNSigmoidLayer to CrossEntropy. CrossEntropy works with (posterior) probability distributions and the input must sum up to 1. The output of DNNSigmoidLayer(it applies sigmoid, not softmax) will not be normalized to sum up to 1.

You need to normalize the output of `sigmoid `(since you have multi-class problem) if you want to use `CrossEntropy`, which is why I recommended using `CrossEntropyWithSoftmax` directly. 

Did you try using the code I sent you? Could you share what was the result?

Finally, I am looking at the code excerpt from you last post. In the last line, did you really mean to use
`ce = CrossEntropy (ageol, h3)` or did you want to use `age` instead of `ageol` in there?

**Comment by spandantiwari** [comment]
Chhyun - any updates?

**Comment by chhyun** [comment]
Sorry, I was busy for conference and just came back yesterday.

And now I'm working on my graduation thesis.

So I'll be busy for a while...

After I submit my graduation thesis, I'll try again and notice you about
the results.

2017-06-21 6:09 GMT+09:00 Spandan Tiwari <notifications@github.com>:

**Comment by spandantiwari** [comment]
Sure. Please re-open the issue if you have questions after trying it out.
